lib/puppet/functions/reverse_each.rb

```diff
@@ -0,0 +1,88 @@
+# Reverses the order of the elements of something that is iterable and optionally runs a
+# [lambda](http://docs.puppetlabs.com/puppet/latest/reference/lang_lambdas.html) for each
+# element.
+#
+# This function takes one to two arguments.
+#
+# 1. An iterable that the function will iterate over.
+# 2. An optional lambda, which the function calls for each element in the first argument. It must
+# request one parameter.
+#
+# @example Using the `reverse_each` function
+#
+# `$data.reverse_each |$parameter| { <PUPPET CODE BLOCK> }`
+#
+# or
+#
+# `$reverse_data = $data.reverse_each`
+#
+# or
+#
+# `reverse_each($data) |$parameter| { <PUPPET CODE BLOCK> }`
+#
+# or
+#
+# `$reverse_data = reverse_each($data)`
+#
+# When no second argument is present, Puppet returns an iterable that represents the reverse
+# order of its first argument. This allows methods on iterables to be chained.
+#
+# When a lamdba is given as the second argument, Puppet iterates the first argument in reverse order
+# and passes each value in turn to the lambda, then returns the first argument unchanged.
+#
+# @example Using the `reverse_each` function with an array and a one-parameter lambda
+#
+# ~~~ puppet
+# # For the array $data, run a lambda that creates a resource in reverse order for each item in reverse.
+# $data = ["routers", "servers", "workstations"]
+# $data.reverse_each |$item| {
+#  notify { $item:
+#    message => $item
```

Comment:
Also, adding resources in a particular order is kind of meaningless since they are lazily evaluated. It best if the example shows something where order actually matters.